packages/actor-abstract-parse/README.md

```diff
@@ -0,0 +1,19 @@
+# Comunica Abstract Parse Actor
```

Comment:
Hopefully this addresses several comments:

The `actor-abstract-parse` takes a `NodeJS.Readable`, and `metadata` (e.g. a `baseIRI`) as input, and outputs a stream of parsed data. This can then be instantiated (for instance in `bus-rdf-parse`) with a particular type for the output stream and input/output metadata (e.g. `RDF.Stream`). This has *nothing* to do with the dereferencing.

The `actor-abstract-dereference-parse` then uses two mediators - the `bus-dereference` mediator, and *any* mediator that satisfies the `actor-abstract-parse` interface (e.g. `bus-rdf-parse`, `bus-shacl-parse`, `bus-rule-parse` etc.). It is instantiated as a `dereference-parse` bus (e.g. `bus-dereference-rdf`, `bus-dereference-shacl` or `bus-dereference-rule`). This allows us to wire our dereference mediator  to any parse mediator.

The `bus-dereference` does not handle any parsing; it simply takes a url/file-path and outputs a NodeJS.ReadableStream.

Reply:
There are a couple of options to change things here:

If we had something like 'actor-groups' mentioned in https://github.com/comunica/comunica/issues/907 then we could just have a `bus-dereference-parse` and then have some configs that look like

rdf-dereference/actors.json
```json
{
  "@context": [
    "https://linkedsoftwaredependencies.org/bundles/npm/@comunica/runner/^2.0.0/components/context.jsonld",

    "https://linkedsoftwaredependencies.org/bundles/npm/@comunica/bus-dereference-parse/^2.0.0/components/context.jsonld"
  ],
  "@id": "urn:comunica:default:Runner",
  "@type": "Runner",
  "actors": [
    {
      "@id": "urn:comunica:default:dereference-parse/actors#rdf",
      "@type": "ActorDereferenceParse",
      "ex:group": "urn:comunica:default:dereference-parse/actors#rdf",
      "mediatorParseMediatypes": { "@id": "urn:comunica:default:rdf-parse/mediators#mediaType" },
      "mediatorParse": { "@id": "urn:comunica:default:rdf-parse/mediators#parse" },
      "mediatorDereference": { "@id": "urn:comunica:default:dereference/mediators#main" }
    }
  ]
}
```

rdf-dereference/mediator.json
```json
{
  "@context": [
    "https://linkedsoftwaredependencies.org/bundles/npm/@comunica/bus-rdf-dereference/^2.0.0/components/context.jsonld",
    "https://linkedsoftwaredependencies.org/bundles/npm/@comunica/mediator-race/^2.0.0/components/context.jsonld"
  ],
  "@id": "urn:comunica:default:rdf-dereference/mediators#main",
  "@type": "MediatorRace",
  "bus": { "@id": "ActorRdfDereference:_default_bus" }
  "ex:group": "urn:comunica:default:dereference-parse/actors#rdf",
}
```

and then so on for shacl-dereference etc,

Reply:
Ok, got it.
I see now how all components are linked to each other (and are needed indeed).

Since most (all?) dereference implementations will probably require some kind of parsing, would it make sense to move the abstract dereference parse actor into the dereference bus? We could then also even move `IActionParse, IActorParseOutput, IParseMetadata` to the dereference bus as well (`ActorAbstractParse` is not needed even?).
This would remove the need for these two additional packages, and may increase understandability.

Reply:
> Since most (all?) dereference implementations will probably require some kind of parsing, would it make sense to move the abstract dereference parse actor into the dereference bus?

Perhaps I tried to 'over-modularise' things here, and I was thinking something along these lines at one point - but I still have a couple of reservations:


If I understand correctly - this means that the `bus-dereference` would have a parameter which is a callback to a parser; then `bus-dereference-rdf`, `bus-dereference-shacl` etc. would look something like

```ts
run(action) {
   return this.mediatorDereference.mediate({ action, parser: this.mediatorRdfParse.mediate })
}
```

Hence each `bus-dereference-x` actor is going to need to repeat this logic of invoking the dereference mediator and giving it the relevant callback.

> We could then also even move IActionParse, IActorParseOutput, IParseMetadata to the dereference bus as well (ActorAbstractParse is not needed even?).

My concern about this is that it would mean that the `bus-rdf-parse`, `bus-shacl-parse` etc. module would then have a dependency on the `bus-dereference` package for those types even though they need not actually interact with the dereference bus for certain applications (e.g. a configuration that creates a parser module which accepts a `NodeJS.Readable` stream as input).

Reply:
> If I understand correctly - this means that the bus-dereference would have a parameter which is a callback to a parser; then bus-dereference-rdf, bus-dereference-shacl etc. would look something like

Ah no, that's not what I mean. Your current implementation of `AbstractDereferenceParse` could still be kept as-is (aside from some renamings perhaps).

The `bus-dereference` would just have two base actors now: `ActorDereference` (=the base), and `ActorDereferenceParse` (=the `AbstractDereferenceParse` you have now, but perhaps also extending from `ActorDereference`).

> My concern about this is that it would mean that the bus-rdf-parse, bus-shacl-parse etc. module would then have a dependency on the bus-dereference package

Right, good point. Then keeping that package might be the cleanest option indeed. I would still remove `ActorAbstractParse` though, as it's a no-op atm.